### PR TITLE
Feature/issue 5 current article exposed as json

### LINKED ISSUE #5: current article exposed as json
Expose the article that is "published" and has a publish date of less than 7 days past in json format. No authentication is required.

Data should be literally as simple as:

```
{
    "article": "this is an article up to 480 chars long"
}
```

If no article is published return:

```
{
     "article": "Sorry there's no article this week, dial back soon!"
}
```

Ensure data is json encoded and returns correct content-type. See [here](http://stackoverflow.com/a/13011046/1549561) for details.

### PR DESCRIPTION
**Pull request by gitumarkk**
Fix #5 
- Created articles API in article/views.py
- Created unittests in a test_views.py
- Added url resource to url.py

### CODE REVIEW

File: articles/urls.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+from django.conf.urls import patterns, url
+
+urlpatterns = patterns('',
+    url(r'^articles/$', 'articles.views.get_articles'),
```
Comment (imsickofmaps) [inline articles/urls.py]: drop the /  and make all references singular as we're always working with one article

**Comment by imsickofmaps** [comment]
@smn ready for your eyes - made my comments

File: articles/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        Article.objects.create(article="Test4",
+                               publish=True,
+                               publish_at=
+                               (timezone.now() + datetime.timedelta(days=3)))
+
+        Article.objects.create(article="Test5",
+                               publish=False,
+                               publish_at=
+                               (timezone.now() - datetime.timedelta(days=2)))
```
Comment (smn) [inline articles/tests/test_views.py]: fixtures will probably make this easier and more reusable?

File: articles/tests/test_views.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        response = self.client.get("/api/article")
+        self.assertEqual("application/json", response["Content-Type"])
+        self.assertIn('Test2', response.content)
```
Comment (smn) [inline articles/tests/test_views.py]: You're getting back JSON, a better test is to do `json.loads(response.content)` which'll give you back a dictionary and do asserts on that?

File: articles/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                               publish=True,
+                               publish_at=
+                               (timezone.now() + datetime.timedelta(days=3)))
+
+        Article.objects.create(article="Test5",
+                               publish=False,
+                               publish_at=
+                               (timezone.now() - datetime.timedelta(days=2)))
+
+        response = self.client.get("/api/article")
```
Comment (smn) [inline articles/tests/test_views.py]: Please use Django's `reverse()` url helpers to avoid hard coding URLs anywhere.

File: articles/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                               publish_at=
+                               (timezone.now() - datetime.timedelta(days=8)))
+
+        Article.objects.create(article="Test7",
+                               publish=True,
+                               publish_at=
+                               (timezone.now() + datetime.timedelta(seconds=60)))
+
+        response = self.client.get("/api/article")
+        self.assertIn("application/json", response.items()[0])
```
Comment (smn) [inline articles/tests/test_views.py]: check for `response['Content-Type']`, the order of headers in an HTTP response isn't guaranteed (or important).

File: articles/tests/test_views.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        self.assertIn("Sorry there's no article this week", response.content)
```
Comment (smn) [inline articles/tests/test_views.py]: test for json

File: articles/urls.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+from django.conf.urls import patterns, url
+
+urlpatterns = patterns('',
+    url(r'^article$', 'articles.views.get_article'),
```
Comment (smn) [inline articles/urls.py]: I know @imsickofmaps advised to remove the trailing slash, I'm suggesting you add it back in. Django by default ships with a middleware that [appends slashes](https://docs.djangoproject.com/en/dev/ref/settings/#append-slash) if a page 404s to see if it does exist with a trailing slash. This is a Django convention we follow and has to do with keeping the paths of relative paths more predictable.

File: articles/views.py [hunk 8]
```diff
@@ -1 +1,22 @@
...
+
+
+def get_article(request):
+    timedelta = timezone.now() - datetime.timedelta(days=7)
+    try:
+        queryset = (Article.objects.all().
+                    filter(publish=True).
+                    filter(publish_at__lte=timezone.now()).
+                    filter(publish_at__gte=timedelta).
+                    order_by('publish_at')[0])
```
Comment (smn) [inline articles/views.py]: You want to use [`.exists()`](https://docs.djangoproject.com/en/1.5/ref/models/querysets/#django.db.models.query.QuerySet.exists) here, it is more efficient and prevents the need for the IndexError.

File: articles/views.py [hunk 9]
```diff
@@ -1 +1,22 @@
-# Create your views here.
+from django.http import HttpResponse
+from articles.models import Article
+from django.utils import simplejson
```
Comment (smn) [inline articles/views.py]: We're using Python 2.6, it ships with the `json` module, rather use that than `simplejson` as the latter sometimes returns bytecode strings and sometimes unicode. It does this because of performance reasons, we'd rather take the performance hit for predictability.

File: articles/views.py [hunk 10]
```diff
@@ -1 +1,22 @@
...
+                    filter(publish=True).
+                    filter(publish_at__lte=timezone.now()).
+                    filter(publish_at__gte=timedelta).
+                    order_by('publish_at')[0])
+        output = {"article": queryset.article}
+    except IndexError:
+        output = {"article":
+                  "Sorry there's no article this week, dial back soon!"}
+
+    return HttpResponse(simplejson.dumps(output),
```
Comment (smn) [inline articles/views.py]: Before you guys invent your own rest-ful API stuff, have a look at https://github.com/toastdriven/django-tastypie.

File: articles/tests/test_views.py (hunk 4 above)
Comment (gitumarkk) [inline articles/tests/test_views.py]: Ok

File: articles/tests/test_views.py (hunk 3 above)
Comment (gitumarkk) [inline articles/tests/test_views.py]: Ok

File: articles/tests/test_views.py (hunk 5 above)
Comment (gitumarkk) [inline articles/tests/test_views.py]: A later commit [https://github.com/praekelt/go-nike-ghr/commit/b909e962d6593b838fde5e603c82f216c0ba1a27#L0R38] sorts this issue out.

File: articles/tests/test_views.py (hunk 6 above)
Comment (gitumarkk) [inline articles/tests/test_views.py]: Ok

File: articles/urls.py (hunk 7 above)
Comment (gitumarkk) [inline articles/urls.py]: Cool, will revert back

File: articles/views.py (hunk 8 above)
Comment (gitumarkk) [inline articles/views.py]: Will change it.

File: articles/tests/test_views.py (hunk 5 above)
Comment (imsickofmaps) [inline articles/tests/test_views.py]: @gitumarkk - no it doesn't, otherwise @smn wouldn't see it in the diff! This is a second reference to the same issue. :smile:

File: articles/views.py (hunk 10 above)
Comment (imsickofmaps) [inline articles/views.py]: @gitumarkk please follow the advice here - I've discussed on IRC with @smn and he's right. As usual.

Comment (gitumarkk) [inline articles/views.py]: @imsickofmaps shall I use django-tastypie or the django-rest-framework as was in the Django Text book

Comment (imsickofmaps) [inline articles/views.py]: @gitumarkk use tastypie

File: ghr/urls.py [hunk 11]
```diff
@@ -11,8 +11,10 @@
 
     # Uncomment the admin/doc line below to enable admin documentation:
     # url(r'^admin/doc/', include('django.contrib.admindocs.urls')),
-
     # Uncomment the next line to enable the admin:
     url(r'^admin/', include(admin.site.urls)),
     url(r'^celery/', include('celery_app.urls')),
+
+    # API links
+    url(r'^api/', include('articles.urls')),
```
Comment (imsickofmaps) [inline ghr/urls.py]: Please make sure we structure the URLs for the API in a way that means when the next application gets added this still works.

**Comment by smn** [comment]
@imsickofmaps @gitumarkk Is this ready for re-review?

**Comment by imsickofmaps** [comment]
@smn I've asked for a some comments to be added to the code but if you have a moment to review without them then as soon as they're added I'll merge

**Comment by gitumarkk** [comment]
@imsickofmaps added comments and made latest commit of the code.

**Comment by imsickofmaps** [comment]
@smn Ready for review. Hopefully last.

**Comment by imsickofmaps** [comment]
@smn really ready. :grin:

File: articles/admin.py [hunk 12]
```diff
@@ -1,4 +1,21 @@
...
+        # Overiding the page CharField widget to Text Area
+        formfield = super(ArticleAdmin, self).formfield_for_dbfield(db_field, **kwargs)
+        if db_field.name == 'page_1':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
+        if db_field.name == 'page_2':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
+        if db_field.name == 'page_3':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
+        if db_field.name == 'page_4':
+            formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
```
Comment (smn) [inline articles/admin.py:12]: This seems like a long way around for either (ModelAdmin.formfield_overrides](https://docs.djangoproject.com/en/1.5/ref/contrib/admin/#django.contrib.admin.ModelAdmin.formfield_overrides) or specifying a custom (ModelAdmin.form)[https://docs.djangoproject.com/en/dev/ref/contrib/admin/#django.contrib.admin.ModelAdmin.form] that sets the `widget=forms.TextArea` on the form fields.

If anything else then reduce this to:

``` python
if db_field.name in ['page_1','page_2','page_3','page_4']:
    formfield.widget = forms.Textarea(attrs=formfield.widget.attrs)
```

File: articles/api.py [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    def get_object_list(self, request):
+        # Filters the queryset in meta to get the specific Article required
+        timedelta = timezone.now() - datetime.timedelta(days=7)
+        query = super(ArticleResource, self).get_object_list(request)
+        query = (query.filter(publish=True).
+                 filter(publish_at__lte=timezone.now).
+                 filter(publish_at__gte=timedelta).
+                 order_by('publish_at'))
+
+        if query.exists():
```
Comment (smn) [inline articles/api.py]: why is this necessary? Is returning a `query` that doesn't have any results not the same as returning `Article.objects.none()` ?

File: articles/api.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            return query
+        else:
+            # Need to create an empty object otherwise tastypie complains
+            return Article.objects.none()
+
+    def alter_list_data_to_serialize(self, request, data_dict):
```
Comment (smn) [inline articles/api.py:29]: My gut feel is that it'd be easier to leave this an API that just returns whatever it has and leaves the JS application to make sense of it.

File: articles/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+                      'api_name': 'api'})
+        response = self.client.get(url)
+        self.assertEqual("application/json", response["Content-Type"])
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(json.loads(response.content)['article'][0],
+                         unicode('Test 1'))
+
+    def test_correct_url(self):
+        # Tests teh URL path to see if it is what its expected
+        response = self.client.get("/api/article/")
```
Comment (smn) [inline articles/tests/test_views.py]: use `reverse()` please

**Comment by smn** [comment]
:+1: other than my small-ish comments.

**Comment by smn** [comment]
my comment with the `alter_list_data_to_serialize()` hook is certainly up for debate.

**Comment by gitumarkk** [comment]
@imsickofmaps made commit
